design/expected-hardware-configuration-validation.md

```diff
@@ -0,0 +1,199 @@
+https://github.com/metal3-io/baremetal-operator/blob/master/deploy/crds/metal3.io_baremetalhosts_crd.yaml
+
+* Write a below schema for new CRD under folder deploy/crds for Kind HardwareValidator.
+
+    ```yaml
+    ExpectedHardwareConfiguration:
+        default:
+        properties:
+           cpu:
+               properties:
+                   count:
+                       type : Integer
+           type: Object
+           disk:
+               properties:
+                   sizeBytesGB:
+                       type: Integer
+                   numberOfDisks:
+                       type: Integer
+           type: Object
+           nics:
+               properties:
+                  numberofNICS:
+                     type: Integer
+           type: Object
+           ram:
+               properties:
+                   sizeBytesGB:
+                       type: Integer
+           type: Object
+           systemVendor:
+              properties:
+                  name:
+                      type: String
+           type: Object
+           Firmware:
+              Properties:
+                  Version:
+                      RAID:
+                          type: String
+                      BasebandManagement:
+                          type: String
+                      BIOS: 
+                          type: String
+                     IDRAC: 
+                          type: String
+            type: Object
+        required:
+        - cpu
+        - nics
+        - ram
+        - disk
+    ```
+* Write a new API as baremetalhost/v1beta1 and new kind(CRD) HardwareValidator.
+    e.g.
+        
+       kubebuilder create api --group baremetalhost --version v1beta1 --kind HardwareValidator
+    
+    - This will create the files api/v1beta1/hardwarevalidator_types.go where the API is defined and the
+      controller/hardwarevalidator_controller.go where the reconciliation business logic is implemented for this Kind(CRD).
+    - Implement a new function fetchHost() which will fetch all baremetal hosts.
+    - In hardwarevalidator_controller.go, reconcile function will call fetchHost() function to fetch all baremetal hosts and also extract
+      Expected hardware configuration from `metal3.io_hardwarevalidator_crd.yaml`.
+    
+        Create a new Validator.go file to write comparison and validation logic for provisioned baremetal hosts.
+        - Write a function which will have the expected hardware details and all the baremetal host list.
+        - Will pass above two inputs to validator function defined in validator.go file.
+	    - Write an algorithm to loop over all the hosts and check for comparison and validation of
+	    specs against the expected hardware details.
+	    - If host match found after execution of above algorithm, matched host will append to list.
+        - Return list to caller function.
+   
+    - According to list returned by validator function, will update the label for all hosts.
+
```

Comment:
Yes, @dhellmann. For user point of view, adding multiple profils labels in host will be nice idea. We will change our implementation as we will check one host against multiple validator profile at a time. Also we will add a watch to the host list for the hardware setting changes and will update the labels accordingly.

Reply:
This section is still talking about working on lists of hosts. I think that's going to lead to a lot of reconcile conflicts.

Instead of a prose version of the algorithm, let's try to write down what conditions will lead the controller to make different types of changes.

For example, what causes the controller to update a host (which CRDs does it watch)? I think it needs to watch both, and I think Reconcile() should work on one BareMetalHost at a time.

What happens if a host used to match a classification profile but no longer does? Do we remove the label or leave it there? I think it's better to remove the label, but will it have any side-effects that consumers of the labels need to be aware of?